<a href="https://colab.research.google.com/github/IvRodco/EstadisticaAplicasa_1899843/blob/main/Actividad6_ILRA_1899843_EAG41_sep2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Estadistica aplicada

##Actividad 6 Parte A

###Ivanna LIzeth Rodriguez Ambriz
###1899843
Grupo 41

In [32]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

In [ ]:
pip install lifelines

In [34]:
from lifelines import CoxPHFitter
import seaborn as sns

In [81]:
df_empleados= pd.read_csv('https://raw.githubusercontent.com/jimmyzac/Estadistica-Aplicada-FCFM-UANL/main/bases_datos/desercion_laboral.csv')
df_empleados

satisfaccion  evaluacion  proyectos  promedio_horas_mes  semestre  \
0               3.8         5.3          2                 157         3   
1               8.0         8.6          5                 262         6   
2               1.1         8.8          7                 272         4   
3               7.2         8.7          5                 223         5   
4               3.7         5.2          2                 159         3   
...             ...         ...        ...                 ...       ...   
14994           4.0         5.7          2                 151         3   
14995           3.7         4.8          2                 160         3   
14996           3.7         5.3          2                 143         3   
14997           1.1         9.6          6                 280         4   
14998           3.7         5.2          2                 158         3   

       accidente  renunciar  ascenso_5years departamento salario  
0              0          1               0       ventas    bajo  
1              0          1               0       ventas   medio  
2              0          1               0       ventas   medio  
3              0          1               0       ventas    bajo  
4              0          1               0       ventas    bajo  
...          ...        ...             ...          ...     ...  
14994          0          1               0      soporte    bajo  
14995          0          1               0      soporte    bajo  
14996          0          1               0      soporte    bajo  
14997          0          1               0      soporte    bajo  
14998          0          1               0      soporte    bajo  

[14999 rows x 10 columns]

In [82]:
#vamos a verificar que no existan missing values

#revisamos cuantos missing values tiene cada columna
df_empleados.isnull().sum()

satisfaccion           0
evaluacion             0
proyectos              0
promedio_horas_mes     0
semestre               0
accidente              0
renunciar              0
ascenso_5years         0
departamento          26
salario                0
dtype: int64

In [83]:
#No nos sirve la variable departamento entonces la quitamos
df_empleados=df_empleados.drop('departamento',axis=1)

In [84]:
#imputamos valores/los eliminamos de la base

#si los eliminamos revisamos cuantos datos perderemos
df_empleados.count()

satisfaccion          14999
evaluacion            14999
proyectos             14999
promedio_horas_mes    14999
semestre              14999
accidente             14999
renunciar             14999
ascenso_5years        14999
salario               14999
dtype: int64

In [85]:
#volvemos a revisar que ya no haya na

df_empleados.isnull().sum()


satisfaccion          0
evaluacion            0
proyectos             0
promedio_horas_mes    0
semestre              0
accidente             0
renunciar             0
ascenso_5years        0
salario               0
dtype: int64

In [86]:
df_empleados.dtypes

satisfaccion          float64
evaluacion            float64
proyectos               int64
promedio_horas_mes      int64
semestre                int64
accidente               int64
renunciar               int64
ascenso_5years          int64
salario                object
dtype: object

In [87]:
#Hacemos los dummies
#Modelo de Cox (en libreria lifelines) solo admite variables continuas y variables dummies(0,1)

#En la libreria lifelines las variables categoricas se tienen que convertir en dummies (con pandas)

dummies=pd.get_dummies(df_empleados['salario'], prefix='salario').astype(int)


In [88]:
dummies.describe()

salario_alto  salario_bajo  salario_medio
count  14999.000000  14999.000000   14999.000000
mean       0.082472      0.487766       0.429762
std        0.275092      0.499867       0.495059
min        0.000000      0.000000       0.000000
25%        0.000000      0.000000       0.000000
50%        0.000000      0.000000       0.000000
75%        0.000000      1.000000       1.000000
max        1.000000      1.000000       1.000000

In [89]:
#solo nos quedamos con 2 dummies
dummies=dummies[['salario_alto','salario_bajo']]


In [90]:
#unimos dummies con df_empleados

df_empleados=pd.concat([df_empleados,dummies],axis=1)
df_empleados=df_empleados.drop('salario',axis=1)

In [91]:
#corroborar que sean numericas que que no haya valores perdidos

df_empleados.dtypes

satisfaccion          float64
evaluacion            float64
proyectos               int64
promedio_horas_mes      int64
semestre                int64
accidente               int64
renunciar               int64
ascenso_5years          int64
salario_alto            int64
salario_bajo            int64
dtype: object

In [95]:
df_empleados.isnull()

satisfaccion  evaluacion  proyectos  promedio_horas_mes  semestre  \
0             False       False      False               False     False   
1             False       False      False               False     False   
2             False       False      False               False     False   
3             False       False      False               False     False   
4             False       False      False               False     False   
...             ...         ...        ...                 ...       ...   
14994         False       False      False               False     False   
14995         False       False      False               False     False   
14996         False       False      False               False     False   
14997         False       False      False               False     False   
14998         False       False      False               False     False   

       accidente  renunciar  ascenso_5years  salario_alto  salario_bajo  
0          False      False           False         False         False  
1          False      False           False         False         False  
2          False      False           False         False         False  
3          False      False           False         False         False  
4          False      False           False         False         False  
...          ...        ...             ...           ...           ...  
14994      False      False           False         False         False  
14995      False      False           False         False         False  
14996      False      False           False         False         False  
14997      False      False           False         False         False  
14998      False      False           False         False         False  

[14999 rows x 10 columns]

In [96]:
#estadisticas descriptivas

df_empleados.describe()

satisfaccion    evaluacion     proyectos  promedio_horas_mes  \
count  14999.000000  14999.000000  14999.000000        14999.000000   
mean       6.128335      7.161017      3.803054          201.050337   
std        2.486307      1.711691      1.232592           49.943099   
min        0.900000      3.600000      2.000000           96.000000   
25%        4.400000      5.600000      3.000000          156.000000   
50%        6.400000      7.200000      4.000000          200.000000   
75%        8.200000      8.700000      5.000000          245.000000   
max       10.000000     10.000000      7.000000          310.000000   

           semestre     accidente     renunciar  ascenso_5years  salario_alto  \
count  14999.000000  14999.000000  14999.000000    14999.000000  14999.000000   
mean       3.498233      0.144610      0.238083        0.021268      0.082472   
std        1.460136      0.351719      0.425924        0.144281      0.275092   
min        2.000000      0.000000      0.000000        0.000000      0.000000   
25%        3.000000      0.000000      0.000000        0.000000      0.000000   
50%        3.000000      0.000000      0.000000        0.000000      0.000000   
75%        4.000000      0.000000      0.000000        0.000000      0.000000   
max       10.000000      1.000000      1.000000        1.000000      1.000000   

       salario_bajo  
count  14999.000000  
mean       0.487766  
std        0.499867  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        1.000000

In [97]:
df_empleados['renunciar'].value_counts()

#1 si renuncio
#0 dato censurado

0    11428
1     3571
Name: renunciar, dtype: int64

##ESTADISTICAS DESCRIPTIVAS

El 23.8% ha renunciado

El promedio de satisfaccion de haber trabajado en la empresa es del 6.1/10

Los empleados trabajan un promedio de 201 horas en el mes

La probabilidad de sufrir un accidente en la empresa es del 14.46%

El 2% de los empleados han sido ascendidos en los ultimos 5 años

El 8.34% de los empleados tienen un salario alto, mientras que el 48.77% tiene un salario bajo



In [102]:
cph1=CoxPHFitter().fit(df_empleados,'satisfaccion','renunciar')
cph1.print_summary()

<lifelines.CoxPHFitter: fitted with 14999 total observations, 11428 right-censored observations>
             duration col = 'satisfaccion'
                event col = 'renunciar'
      baseline estimation = breslow
   number of observations = 14999
number of events observed = 3571
   partial log-likelihood = -31895.52
         time fit was run = 2023-09-27 01:34:01 UTC

---
                     coef  exp(coef)   se(coef)   coef lower 95%   coef upper 95%  exp(coef) lower 95%  exp(coef) upper 95%
covariate                                                                                                                  
evaluacion          -0.09       0.91       0.01            -0.11            -0.07                 0.89                 0.93
proyectos            0.06       1.07       0.02             0.03             0.10                 1.03                 1.10
promedio_horas_mes   0.00       1.00       0.00             0.00             0.00                 1.00                 1.00
semestre             0.18       1.19       0.01             0.16             0.20                 1.17                 1.22
accidente           -1.32       0.27       0.08            -1.47            -1.16                 0.23                 0.31
ascenso_5years      -1.26       0.28       0.23            -1.71            -0.81                 0.18                 0.44
salario_alto        -1.17       0.31       0.11            -1.40            -0.95                 0.25                 0.39
salario_bajo         0.42       1.53       0.04             0.35             0.49                 1.42                 1.63

                     cmp to      z      p   -log2(p)
covariate                                           
evaluacion             0.00  -8.80 <0.005      59.31
proyectos              0.00   3.86 <0.005      13.08
promedio_horas_mes     0.00   6.38 <0.005      32.37
semestre               0.00  18.81 <0.005     259.82
accidente              0.00 -16.68 <0.005     205.07
ascenso_5years         0.00  -5.47 <0.005      24.43
salario_alto           0.00 -10.30 <0.005      80.21
salario_bajo           0.00  12.05 <0.005     108.69
---
Concordance = 0.67
Partial AIC = 63807.04
log-likelihood ratio test = 1269.63 on 8 df
-log2(p) of ll-ratio test = 890.50

In [118]:
(1-0.95)*100

5.000000000000004

H0:Afecta a la satisfaccion dentro de la empresa y por tanto la proporcion de empleados que renuncian

H1:No afecta a la satisfaccion dentro de la empresa y por tanto la proporcion de empleados que renuncian

#Interpretacion de coeficientes

**evaluacion**: No fecta a la proporcion de empleados que renuncian en un 8.9%

**proyectos**:Afecta a la proporcion de empleados que renuncian en un 7%

**promedio_horas_mes**: Afecta a la proporcion de empleados que renuncian

**semestre**: Afecta a la proporcion de empleados que renuncian en un 18%

**accidente**: No afecta a la proporcion de empleados que renuncian en un 73%

**ascenso_5years**:No afecta a la proporcion de empleados que renuncian en un 72%

**salario_alto**: No afecta a la proporcion de empleados que renuncian en un 69%

**salario_bajo**: Afecta a la proporcion de empleados que renuncian en un 53%

In [113]:
#vamos a comprobar el supuesto
#H0:se viola el supuesto
#H1:se cumple el supuesto
cph1.check_assumptions(df_empleados,p_value_threshold=0.05)


The ``p_value_threshold`` is set at 0.05. Even under the null hypothesis of no violations, some
covariates will be below the threshold by chance. This is compounded when there are many covariates.
Similarly, when there are lots of observations, even minor deviances from the proportional hazard
assumption will be flagged.

With that in mind, it's best to use a combination of statistical tests and visual tests to determine
the most serious violations. Produce visual plots using ``check_assumptions(..., show_plots=True)``
and looking for non-constant lines. See link [A] below for a full example.



<lifelines.StatisticalResult: proportional_hazard_test>
 null_distribution = chi squared
degrees_of_freedom = 1
             model = <lifelines.CoxPHFitter: fitted with 14999 total observations, 11428 right-censored observations>
         test_name = proportional_hazard_test

---
                         test_statistic      p  -log2(p)
accidente          km              0.09   0.77      0.38
                   rank            0.02   0.88      0.19
ascenso_5years     km              0.12   0.73      0.46
                   rank            0.46   0.50      1.01
evaluacion         km            243.61 <0.005    180.02
                   rank          114.26 <0.005     86.18
promedio_horas_mes km             16.32 <0.005     14.19
                   rank          109.80 <0.005     82.93
proyectos          km            480.75 <0.005    351.57
                   rank         1047.80 <0.005    761.17
salario_alto       km              0.22   0.64      0.64
                   rank            0.00   1.00      0.01
salario_bajo       km              1.27   0.26      1.95
                   rank            2.19   0.14      2.85
semestre           km            235.89 <0.005    174.43
                   rank          221.72 <0.005    164.17



1. Variable 'evaluacion' failed the non-proportional test: p-value is <5e-05.

   Advice 1: the functional form of the variable 'evaluacion' might be incorrect. That is, there may
be non-linear terms missing. The proportional hazard test used is very sensitive to incorrect
functional forms. See documentation in link [D] below on how to specify a functional form.

   Advice 2: try binning the variable 'evaluacion' using pd.cut, and then specify it in
`strata=['evaluacion', ...]` in the call in `.fit`. See documentation in link [B] below.

   Advice 3: try adding an interaction term with your time variable. See documentation in link [C]
below.


2. Variable 'proyectos' failed the non-proportional test: p-value is <5e-05.

   Advice: with so few unique values (only 6), you can include `strata=['proyectos', ...]` in the
call in `.fit`. See documentation in link [E] below.

3. Variable 'promedio_horas_mes' failed the non-proportional test: p-value is <5e-05.

   Advice 1: the functional f

[]

Dado que los supuestos no cumplen con el riesgo constante, entonces como no se cumple el supuesto debemos utilizar un modelo alternativo, el cual es el Modelo de riesgo acelerado (AFT).

In [114]:
#revisamos la distribucion del riesgo

#importamos la libreria necesaria para esto
from lifelines import ExponentialFitter, WeibullFitter, LogNormalFitter

mexpo= ExponentialFitter().fit(df_empleados['satisfaccion'],df_empleados['evaluacion'])
mweibul=WeibullFitter().fit(df_empleados['satisfaccion'],df_empleados['evaluacion'])
mlogn=LogNormalFitter().fit(df_empleados['satisfaccion'],df_empleados['evaluacion'])

mexpo.AIC_ , mweibul.AIC_ , mlogn.AIC_

(84384.06818356866, 70477.29624807116, 77338.24598274122)

In [117]:
#Dado que con el AIC se elige el de menor valor, eligiremos el Modelo Weibull

#importamos la libreria para el AFT
from lifelines import WeibullAFTFitter

weibulAFT=WeibullAFTFitter().fit(df_empleados,'satisfaccion','renunciar')
weibulAFT.print_summary()

<lifelines.WeibullAFTFitter: fitted with 14999 total observations, 11428 right-censored observations>
             duration col = 'satisfaccion'
                event col = 'renunciar'
   number of observations = 14999
number of events observed = 3571
           log-likelihood = -14325.95
         time fit was run = 2023-09-27 02:06:20 UTC

---
                             coef  exp(coef)   se(coef)   coef lower 95%   coef upper 95%  exp(coef) lower 95%  exp(coef) upper 95%
param   covariate                                                                                                                  
lambda_ accidente            0.94       2.56       0.06             0.83             1.05                 2.28                 2.87
        ascenso_5years       0.90       2.46       0.17             0.58             1.22                 1.78                 3.40
        evaluacion           0.07       1.07       0.01             0.06             0.08                 1.06                 1.09
        promedio_horas_mes  -0.00       1.00       0.00            -0.00            -0.00                 1.00                 1.00
        proyectos           -0.05       0.95       0.01            -0.08            -0.03                 0.93                 0.97
        salario_alto         0.84       2.32       0.08             0.68             1.00                 1.98                 2.73
        salario_bajo        -0.30       0.74       0.03            -0.35            -0.25                 0.70                 0.78
        semestre            -0.13       0.88       0.01            -0.14            -0.11                 0.87                 0.89
        Intercept            3.42      30.66       0.08             3.27             3.58                26.29                35.76
rho_    Intercept            0.33       1.40       0.02             0.30             0.36                 1.36                 1.44

                             cmp to      z      p   -log2(p)
param   covariate                                           
lambda_ accidente              0.00  16.19 <0.005     193.31
        ascenso_5years         0.00   5.44 <0.005      24.16
        evaluacion             0.00   9.37 <0.005      66.86
        promedio_horas_mes     0.00  -6.41 <0.005      32.63
        proyectos              0.00  -4.30 <0.005      15.81
        salario_alto           0.00  10.21 <0.005      78.82
        salario_bajo           0.00 -11.88 <0.005     105.67
        semestre               0.00 -18.16 <0.005     242.47
        Intercept              0.00  43.66 <0.005        inf
rho_    Intercept              0.00  22.12 <0.005     357.62
---
Concordance = 0.67
AIC = 28671.89
log-likelihood ratio test = 1283.29 on 8 df
-log2(p) of ll-ratio test = 900.30

#Interpretacion de coeficientes

**evaluacion**: No afecta a la mediana de empleados que renuncianen un 7%

**proyectos**:No afecta a la mediana de empleados que renuncian en un 5%

**promedio_horas_mes**: No afecta a la mediana de empleados que renuncian

**semestre**: Afecta a la mediana de empleados que renuncian en un

**accidente**: Afecta a la mediana de empleados que renuncian en un 156%

**ascenso_5years**:Afecta a la mediana de empleados que renuncian en un 146%

**salario_alto**: No afecta a la proporcion de empleados que renuncian en un 132%

**salario_bajo**: Afecta a la proporcion de empleados que renuncian en un 36%